From f9cdcb52abec04aab6904c75fb83acd82cf34057 Mon Sep 17 00:00:00 2001
From: Praveen Talari <quic_ptalari@quicinc.com>
Date: Wed, 17 Jul 2024 19:28:52 +0530
Subject: [PATCH] PENDING: i2c: qcom-geni: Enable I2C on SA8255p Qualcomm
 platforms

Enable or disable resources such as clocks, interconnect
paths, and TLMM(GPIOs) via the logical power domain via PM runtime
callbacks.

Change-Id: I067e2af4f8e2d0c89e7ac5d89a34370318d687c3
Signed-off-by: Praveen Talari <quic_ptalari@quicinc.com>
Signed-off-by: Shazad Hussain <quic_shazhuss@quicinc.com>
---
 drivers/i2c/busses/i2c-qcom-geni.c | 239 +++++++++++++++++------------
 1 file changed, 143 insertions(+), 96 deletions(-)

diff --git a/drivers/i2c/busses/i2c-qcom-geni.c b/drivers/i2c/busses/i2c-qcom-geni.c
index ed9e1552b1ff..918ac39b2efe 100644
--- a/drivers/i2c/busses/i2c-qcom-geni.c
+++ b/drivers/i2c/busses/i2c-qcom-geni.c
@@ -92,6 +92,14 @@ enum geni_i2c_err_code {
 #define XFER_TIMEOUT		HZ
 #define RST_TIMEOUT		HZ
 
+struct geni_i2c_desc {
+	bool has_core_clk;
+	char *icc_ddr;
+	bool no_dma_support;
+	unsigned int tx_fifo_depth;
+	int (*geni_se_switch_state)(struct device *dev, bool state);
+};
+
 static int geni_i2c_runtime_resume(struct device *dev);
 
 struct geni_i2c_dev {
@@ -118,16 +126,10 @@ struct geni_i2c_dev {
 	bool gpi_mode;
 	bool abort_done;
 	bool is_shared;
+	const struct geni_i2c_desc *dev_data;
 	unsigned int clk_idx;
 };
 
-struct geni_i2c_desc {
-	bool has_core_clk;
-	char *icc_ddr;
-	bool no_dma_support;
-	unsigned int tx_fifo_depth;
-};
-
 struct geni_i2c_err_log {
 	int err;
 	const char *msg;
@@ -854,6 +856,52 @@ static int setup_gpi_dma(struct geni_i2c_dev *gi2c)
 	return ret;
 }
 
+/*
+ * geni_i2c_switch_resource_state - switch resources from on -> off or off -> on
+ * @dev: pointer to device structure.
+ * @state: decides whether it is on or off.
+ *
+ * return: 0 on success otherwise error.
+ */
+static int geni_i2c_switch_resource_state(struct device *dev, bool state)
+{
+	struct geni_i2c_dev *gi2c = dev_get_drvdata(dev);
+	int ret = 0;
+
+	if (state) {
+		ret = geni_icc_enable(&gi2c->se);
+		if (ret)
+			return ret;
+
+		ret = clk_prepare_enable(gi2c->core_clk);
+		if (ret) {
+			geni_icc_disable(&gi2c->se);
+			return ret;
+		}
+
+		ret = geni_se_resources_on(&gi2c->se);
+		if (ret) {
+			clk_disable_unprepare(gi2c->core_clk);
+			geni_icc_disable(&gi2c->se);
+			return ret;
+		}
+	} else {
+		if (gi2c->is_shared) {
+			geni_se_clks_off(&gi2c->se);
+		} else {
+			ret = geni_se_resources_off(&gi2c->se);
+			if (ret) {
+				enable_irq(gi2c->irq);
+				return ret;
+			}
+		}
+
+		clk_disable_unprepare(gi2c->core_clk);
+		ret = geni_icc_disable(&gi2c->se);
+	}
+	return ret;
+}
+
 static int geni_i2c_probe(struct platform_device *pdev)
 {
 	struct geni_i2c_dev *gi2c;
@@ -874,16 +922,40 @@ static int geni_i2c_probe(struct platform_device *pdev)
 		return PTR_ERR(gi2c->se.base);
 
 	desc = device_get_match_data(&pdev->dev);
+	if (!desc)
+		return -ENXIO;
 
-	if (desc && desc->has_core_clk) {
-		gi2c->core_clk = devm_clk_get(dev, "core");
-		if (IS_ERR(gi2c->core_clk))
-			return PTR_ERR(gi2c->core_clk);
-	}
+	gi2c->dev_data = desc;
 
-	gi2c->se.clk = devm_clk_get(dev, "se");
-	if (IS_ERR(gi2c->se.clk) && !has_acpi_companion(dev))
-		return PTR_ERR(gi2c->se.clk);
+	if (!of_device_is_compatible(dev->of_node, "qcom,sa8255p-geni-i2c")) {
+		if (desc->has_core_clk) {
+			gi2c->core_clk = devm_clk_get(dev, "core");
+			if (IS_ERR(gi2c->core_clk))
+				return PTR_ERR(gi2c->core_clk);
+		}
+
+		gi2c->se.clk = devm_clk_get(dev, "se");
+		if (IS_ERR(gi2c->se.clk) && !has_acpi_companion(dev))
+			return PTR_ERR(gi2c->se.clk);
+
+		ret = geni_icc_get(&gi2c->se, desc->icc_ddr);
+		if (ret)
+			return ret;
+		/*
+		 * Set the bus quota for core and cpu to a reasonable value for
+		 * register access.
+		 * Set quota for DDR based on bus speed.
+		 */
+		gi2c->se.icc_paths[GENI_TO_CORE].avg_bw = GENI_DEFAULT_BW;
+		gi2c->se.icc_paths[CPU_TO_GENI].avg_bw = GENI_DEFAULT_BW;
+
+		if (desc->icc_ddr)
+			gi2c->se.icc_paths[GENI_TO_DDR].avg_bw = Bps_to_icc(gi2c->clk_freq_out);
+
+		ret = geni_icc_set_bw(&gi2c->se);
+		if (ret)
+			return ret;
+	}
 
 	ret = device_property_read_u32(dev, "clock-frequency",
 				       &gi2c->clk_freq_out);
@@ -903,9 +975,11 @@ static int geni_i2c_probe(struct platform_device *pdev)
 	if (ret)
 		return dev_err_probe(dev, ret, "Invalid clk frequency %d Hz\n", gi2c->clk_freq_out);
 
-	ret = geni_se_clk_freq_match(&gi2c->se, I2C_SE_CLK, &gi2c->clk_idx, &clk_rate, true);
-	if (ret)
-		return dev_err_probe(dev, ret, "Failed to find %lu Hz src clk\n", I2C_SE_CLK);
+	if (desc->geni_se_switch_state) {
+		ret = geni_se_clk_freq_match(&gi2c->se, I2C_SE_CLK, &gi2c->clk_idx, &clk_rate, true);
+		if (ret)
+			return dev_err_probe(dev, ret, "Failed to find %lu Hz src clk\n", I2C_SE_CLK);
+	}
 
 	gi2c->adap.algo = &geni_i2c_algo;
 	init_completion(&gi2c->done);
@@ -921,42 +995,28 @@ static int geni_i2c_probe(struct platform_device *pdev)
 	gi2c->adap.dev.of_node = dev->of_node;
 	strscpy(gi2c->adap.name, "Geni-I2C", sizeof(gi2c->adap.name));
 
-	ret = geni_icc_get(&gi2c->se, desc ? desc->icc_ddr : "qup-memory");
-	if (ret)
-		return ret;
-	/*
-	 * Set the bus quota for core and cpu to a reasonable value for
-	 * register access.
-	 * Set quota for DDR based on bus speed.
-	 */
-	gi2c->se.icc_paths[GENI_TO_CORE].avg_bw = GENI_DEFAULT_BW;
-	gi2c->se.icc_paths[CPU_TO_GENI].avg_bw = GENI_DEFAULT_BW;
-	if (!desc || desc->icc_ddr)
-		gi2c->se.icc_paths[GENI_TO_DDR].avg_bw = Bps_to_icc(gi2c->clk_freq_out);
-
-	ret = geni_icc_set_bw(&gi2c->se);
-	if (ret)
-		return ret;
-
-	ret = clk_prepare_enable(gi2c->core_clk);
-	if (ret)
-		return ret;
+	gi2c->suspended = 1;
+	pm_runtime_set_suspended(gi2c->se.dev);
+	pm_runtime_set_autosuspend_delay(gi2c->se.dev, I2C_AUTO_SUSPEND_DELAY);
+	pm_runtime_use_autosuspend(gi2c->se.dev);
+	pm_runtime_enable(gi2c->se.dev);
 
-	ret = geni_se_resources_on(&gi2c->se);
-	if (ret) {
-		dev_err_probe(dev, ret, "Error turning on resources\n");
-		goto err_clk;
+	ret = pm_runtime_get_sync(gi2c->se.dev);
+	if (ret < 0) {
+		dev_err(gi2c->se.dev, "error turning on device :%d\n", ret);
+		goto err;
 	}
+
 	proto = geni_se_read_proto(&gi2c->se);
 	if (proto == GENI_SE_INVALID_PROTO) {
 		ret = geni_load_se_firmware(&gi2c->se, GENI_SE_I2C);
 		if (ret) {
 			dev_err_probe(dev, ret, "i2c firmware load failed ret: %d\n", ret);
-			goto err_resources;
+			goto err;
 		}
 	} else if (proto != GENI_SE_I2C) {
 		ret = dev_err_probe(dev, -ENXIO, "Invalid proto %d\n", proto);
-		goto err_resources;
+		goto err;
 	}
 
 	if (desc && desc->no_dma_support) {
@@ -970,8 +1030,10 @@ static int geni_i2c_probe(struct platform_device *pdev)
 		/* FIFO is disabled, so we can only use GPI DMA */
 		gi2c->gpi_mode = true;
 		ret = setup_gpi_dma(gi2c);
-		if (ret)
-			goto err_resources;
+		if (ret) {
+			dev_err_probe(dev, ret, "Failed to setup GPI DMA mode\n");
+			goto err;
+		}
 
 		I2C_DBG_LOG(LOG_LVL_INFO, dev, "Using GPI DMA mode for I2C\n");
 	} else {
@@ -985,7 +1047,7 @@ static int geni_i2c_probe(struct platform_device *pdev)
 		if (!tx_depth) {
 			ret = dev_err_probe(dev, -EINVAL,
 					    "Invalid TX FIFO depth\n");
-			goto err_resources;
+			goto err;
 		}
 
 		gi2c->tx_wm = tx_depth - 1;
@@ -1001,22 +1063,12 @@ static int geni_i2c_probe(struct platform_device *pdev)
 		I2C_DBG_LOG(LOG_LVL_VERBOSE, &pdev->dev, "Multi-EE usecase with shared SE\n");
 	}
 
-	clk_disable_unprepare(gi2c->core_clk);
-	ret = geni_se_resources_off(&gi2c->se);
-	if (ret) {
-		dev_err_probe(dev, ret, "Error turning off resources\n");
-		goto err_dma;
-	}
+	ret = pm_runtime_put_sync(gi2c->se.dev);
+	if (ret < 0) {
+		dev_err(gi2c->se.dev, "error turning off device :%d\n", ret);
 
-	ret = geni_icc_disable(&gi2c->se);
-	if (ret)
 		goto err_dma;
-
-	gi2c->suspended = 1;
-	pm_runtime_set_suspended(gi2c->se.dev);
-	pm_runtime_set_autosuspend_delay(gi2c->se.dev, I2C_AUTO_SUSPEND_DELAY);
-	pm_runtime_use_autosuspend(gi2c->se.dev);
-	pm_runtime_enable(gi2c->se.dev);
+	}
 
 	ret = i2c_add_adapter(&gi2c->adap);
 	if (ret) {
@@ -1034,16 +1086,12 @@ static int geni_i2c_probe(struct platform_device *pdev)
 
 	return ret;
 
-err_resources:
-	geni_se_resources_off(&gi2c->se);
-err_clk:
-	clk_disable_unprepare(gi2c->core_clk);
-
-	return ret;
-
 err_dma:
 	release_gpi_dma(gi2c);
 	return ret;
+err:
+	pm_runtime_put_sync(gi2c->se.dev);
+	return ret;
 }
 
 static void geni_i2c_remove(struct platform_device *pdev)
@@ -1066,52 +1114,39 @@ static void geni_i2c_shutdown(struct platform_device *pdev)
 
 static int __maybe_unused geni_i2c_runtime_suspend(struct device *dev)
 {
-	int ret;
+	int ret = 0;
 	struct geni_i2c_dev *gi2c = dev_get_drvdata(dev);
 
 	disable_irq(gi2c->irq);
 
-	if (gi2c->is_shared) {
-		geni_se_clks_off(&gi2c->se);
-	} else {
-		ret = geni_se_resources_off(&gi2c->se);
+	if (gi2c->dev_data->geni_se_switch_state) {
+		ret = gi2c->dev_data->geni_se_switch_state(dev, false);
 		if (ret) {
-			enable_irq(gi2c->irq);
+			dev_err(dev, "failed to switch resources, ret:%d\n", ret);
 			return ret;
 		}
 	}
-	gi2c->suspended = 1;
 
-	clk_disable_unprepare(gi2c->core_clk);
-
-	return geni_icc_disable(&gi2c->se);
+	gi2c->suspended = 1;
+	return ret;
 }
 
 static int __maybe_unused geni_i2c_runtime_resume(struct device *dev)
 {
-	int ret;
+	int ret = 0;
 	struct geni_i2c_dev *gi2c = dev_get_drvdata(dev);
 
-	ret = geni_icc_enable(&gi2c->se);
-	if (ret)
-		return ret;
-
-	ret = clk_prepare_enable(gi2c->core_clk);
-	if (ret) {
-		geni_icc_disable(&gi2c->se);
-		return ret;
-	}
-
-	ret = geni_se_resources_on(&gi2c->se);
-	if (ret) {
-		clk_disable_unprepare(gi2c->core_clk);
-		geni_icc_disable(&gi2c->se);
-		return ret;
+	if (gi2c->dev_data->geni_se_switch_state) {
+		ret = gi2c->dev_data->geni_se_switch_state(dev, true);
+		if (ret) {
+			dev_err(dev, "failed to switch resources, ret:%d\n", ret);
+			return ret;
+		}
 	}
 
 	enable_irq(gi2c->irq);
 	gi2c->suspended = 0;
-	return 0;
+	return ret;
 }
 
 static int __maybe_unused geni_i2c_suspend_noirq(struct device *dev)
@@ -1143,16 +1178,28 @@ static const struct dev_pm_ops geni_i2c_pm_ops = {
 									NULL)
 };
 
+static const struct geni_i2c_desc geni_i2c = {
+	.icc_ddr = "qup-memory",
+	.geni_se_switch_state = geni_i2c_switch_resource_state,
+};
+
+static const struct geni_i2c_desc remotely_geni_i2c = {
+	.icc_ddr = "qup-memory",
+	.geni_se_switch_state = NULL,
+};
+
 static const struct geni_i2c_desc i2c_master_hub = {
 	.has_core_clk = true,
 	.icc_ddr = NULL,
 	.no_dma_support = true,
 	.tx_fifo_depth = 16,
+	.geni_se_switch_state = geni_i2c_switch_resource_state,
 };
 
 static const struct of_device_id geni_i2c_dt_match[] = {
-	{ .compatible = "qcom,geni-i2c" },
+	{ .compatible = "qcom,geni-i2c", .data = &geni_i2c },
 	{ .compatible = "qcom,geni-i2c-master-hub", .data = &i2c_master_hub },
+	{ .compatible = "qcom,sa8255p-geni-i2c", .data = &remotely_geni_i2c },
 	{}
 };
 MODULE_DEVICE_TABLE(of, geni_i2c_dt_match);
-- 
2.34.1

